bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/directentrylogger/DirectWriter.java

```diff
@@ -0,0 +1,318 @@
+                try {
+                    int ret = nativeIO.fallocate(fd, NativeIO.FALLOC_FL_ZERO_RANGE, 0, maxFileSize);
+                    checkState(ret == 0, "Exception should have been thrown on non-zero ret: %d", ret);
+                } catch (NativeIOException ex) {
+                    // fallocate(2) is not supported on all filesystems.  Since this is an optimization, disable
+                    // subsequent usage instead of failing the operation.
+                    useFallocate = false;
+                    slog.kv("message", ex.getMessage())
+                        .kv("file", filename)
+                        .kv("errno", ex.getErrno())
+                        .warn(Events.FALLOCATE_NOT_AVAILABLE);
+                }
+            }
+        }
+
+        this.bufferPool = bufferPool;
+        this.nativeBuffer = bufferPool.acquire();
+    }
+
+    @Override
+    public int logId() {
+        return id;
+    }
+
+    @Override
+    public void writeAt(long offset, ByteBuf buf) throws IOException {
+        checkArgument(Buffer.isAligned(offset),
+                      "Offset to writeAt must be aligned to %d: %d is not", Buffer.ALIGNMENT, offset);
+        checkArgument(Buffer.isAligned(buf.readableBytes()),
+                      "Buffer must write multiple of alignment bytes (%d), %d is not",
+                      Buffer.ALIGNMENT, buf.readableBytes());
+        Buffer tmpBuffer = bufferPool.acquire();
+        int bytesToWrite = buf.readableBytes();
+        tmpBuffer.reset();
+        tmpBuffer.writeByteBuf(buf);
+        Future<?> f = writeExecutor.submit(() -> {
+                try {
+                    int ret = nativeIO.pwrite(fd, tmpBuffer.pointer(), bytesToWrite, offset);
+                    if (ret != bytesToWrite) {
+                        throw new IOException(exMsg("Incomplete write")
+                                              .kv("filename", filename)
+                                              .kv("writeSize", bytesToWrite)
+                                              .kv("bytesWritten", ret)
+                                              .kv("offset", offset).toString());
+                    }
+                } catch (NativeIOException ne) {
+                    throw new IOException(exMsg("Write error")
+                                          .kv("filename", filename)
+                                          .kv("writeSize", bytesToWrite)
+                                          .kv("errno", ne.getErrno())
+                                          .kv("offset", offset).toString());
+                } finally {
+                    bufferPool.release(tmpBuffer);
+                }
+                return null;
+            });
+        addOutstandingWrite(f);
+    }
+
+    @Override
+    public int writeDelimited(ByteBuf buf) throws IOException {
+        synchronized (bufferLock) {
+            if (!nativeBuffer.hasSpace(serializedSize(buf))) {
+                flushBuffer();
```

Comment:
The flush operation will be triggered when the nativeBuffer has no remaining space. 

However, if our disk is SSD, it will have better performance for smaller IO than bigger IO.

Our default nativeBuffer size is `0.125 * direct memory / number of ledger directories / 8`, and could be configured by specific parameter.

We'd better to make it to be configured by `flushEntrylogBytes`